include/seqan3/contrib/parallel/reader_writer_latch.hpp

```diff
@@ -0,0 +1,323 @@
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void writer_arrive_and_wait() noexcept
+    {
+        writer_latch.arrive_and_wait();
+
+        std::call_once(flag, completion_fn);
+    }
+
+    /*!\brief Atomically decrements writer counter by one.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for producer threads without blocking the calling thread.
+     * If all participating producer threads arrived at the synchronisation point, one of the
+     * threads will invoke the completion function.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void writer_arrive() noexcept
+    {
+        writer_latch.arrive();
+
+        if (writer_latch.try_wait())
+            std::call_once(flag, completion_fn);
+    }
+
+    /*!\brief Atomically decrements writer counter by one and blocks the calling thread.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for consumer threads and blocks the calling thread until all participating
+     * threads have arrived.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void reader_arrive_and_wait() noexcept
+    {
+        reader_latch.arrive_and_wait();
+    }
+
+    /*!\brief Atomically decrements reader counter by one.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for consumer threads without blocking the calling thread.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void reader_arrive() noexcept
+    {
+        reader_latch.arrive();
+    }
+
+private:
+
+    //!\brief The internal latch for consumer threads.
+    alignas(std::hardware_destructive_interference_size) latch reader_latch;
+    //!\brief The internal latch for producer threads.
+    alignas(std::hardware_destructive_interference_size) latch writer_latch;
+    //!\brief The stored completion function.
+    std::function<void()>                                      completion_fn;
+    //!\brief This flag ensures that the completion phase is invoked only once.
+    std::once_flag                                             flag;
+};
+
+/*!\brief A strictly scope-based seqan3::contrib::reader_writer_latch wrapper for producer threads.
+ *
+ * \details
+ *
+ * The class scoped_writer_guard is a reader_writer_latch wrapper that provides a convenient RAII-style mechanism for
+ * controlling participating producer threads of seqan3::contrib::reader_writer_latch for the duration of a scoped
+ * block. When a scoped_writer_guard object is created, it becomes a participating producer thread of the
+ * seqan3::contrib::reader_writer_latch it is given. When control leaves the scope in which the scoped_writer_guard
```

Comment:
control?

Reply:
I took this from the documentation of https://en.cppreference.com/w/cpp/thread/lock_guard.

Reply:
Okay so control in the sense of control flow.